### ISSUE TITLE
migrate away from ffmpeg_3

### ISSUE THREAD

**Issue by dotlambda** [NixOS/nixpkgs#120705]
`ffmpeg_3` has many open vulnerabilities (see #94003 and #120372). There seems to be no effort to add patches for these, so we should drop `ffmpeg_3` or at least mark it as insecure.
In https://github.com/NixOS/nixpkgs/pull/89264, `ffmpeg_3` was made the de facto default by making every package that depends on `ffmpeg` depend on `ffmpeg_3` instead. I think that was a bad idea given that the Ffmpeg packages aren't well maintained.
Most packages should build just fine with `ffmpeg` but someone needs to test them.

~Is there an easy way to obtain a list of packages using `ffmpeg_3` and ping their maintainers?~

cc @doronbehar @codyopel 

Here's a list of affected packages:

- [x] aegisub (https://github.com/NixOS/nixpkgs/pull/120802)
- [x] airtame (https://github.com/NixOS/nixpkgs/pull/155539)
- [x] ardour (https://github.com/NixOS/nixpkgs/pull/123930)
- [x] attract-mode (https://github.com/NixOS/nixpkgs/pull/126354)
- [x] avxsynth (https://github.com/NixOS/nixpkgs/pull/123478)
- [x] baresip (https://github.com/NixOS/nixpkgs/pull/123480)
- [x] bitwig-studio1 (https://github.com/NixOS/nixpkgs/pull/121283)
- [x] bitwig-studio3 (https://github.com/NixOS/nixpkgs/pull/121283)
- [x] bombono (https://github.com/NixOS/nixpkgs/pull/121289)
- [x] capture (https://github.com/NixOS/nixpkgs/pull/123491)
- [x] cfdg (https://github.com/NixOS/nixpkgs/pull/123340)
- [x] clipgrab (https://github.com/NixOS/nixpkgs/pull/122751)
- [x] deadbeef (https://github.com/NixOS/nixpkgs/pull/123460)
- [x] devede (https://github.com/NixOS/nixpkgs/pull/123465)
- [x] dr14_tmeter (#125241)
- [x] dvd-slideshow (https://github.com/NixOS/nixpkgs/pull/123489)
- [x] ffms (https://github.com/NixOS/nixpkgs/pull/120846)
- [x] gnomecast (https://github.com/NixOS/nixpkgs/pull/122626)
- [x] goldendict (https://github.com/NixOS/nixpkgs/pull/123527)
- [x] gopro (https://github.com/NixOS/nixpkgs/pull/123795)
- [x] grass (@mpickering)
- [x] haskellPackages.cut-the-crap (https://github.com/NixOS/nixpkgs/pull/123350)
- [x] hedgewars (https://github.com/NixOS/nixpkgs/pull/121851)
- [x] oraclejdk (https://github.com/NixOS/nixpkgs/pull/123513)
- [x] k3b (https://github.com/NixOS/nixpkgs/pull/121256)
- [x] kid3 (https://github.com/NixOS/nixpkgs/pull/120802)
- [x] libextractor (https://github.com/NixOS/nixpkgs/pull/123344)
- [x] libgroove (@andrewrk)
- [x] liblinphone (https://github.com/NixOS/nixpkgs/pull/123366)
- [x] libsForQt5.libopenshot (https://github.com/NixOS/nixpkgs/pull/120600)
- [x] libsForQt5.qtwebengine (https://github.com/NixOS/nixpkgs/pull/120646)
- [x] libvdpau-va-gl (https://github.com/NixOS/nixpkgs/pull/123757)
- [x] lightspark (https://github.com/NixOS/nixpkgs/pull/124330)
- [x] linphone (https://github.com/NixOS/nixpkgs/pull/123366)
- [x] makemkv (#121475)
- [x] manim (https://github.com/NixOS/nixpkgs/pull/123799)
- [x] mediastreamer (https://github.com/NixOS/nixpkgs/pull/123366)
- [x] mediatomb (https://github.com/NixOS/nixpkgs/pull/123487)
- [x] megacmd (https://github.com/NixOS/nixpkgs/pull/123711)
- [x] megasync (https://github.com/NixOS/nixpkgs/pull/121236)
- [x] mgba (https://github.com/NixOS/nixpkgs/pull/120802)
- [x] minidlna (https://github.com/NixOS/nixpkgs/pull/121089)
- [x] moc (https://github.com/NixOS/nixpkgs/pull/123486)
- [x] moonlight-embedded (#121849)
- [x] musikcube (https://github.com/NixOS/nixpkgs/pull/121285)
- [x] natron (@puffnfresh)
- [x] nginxModules.video-thumbextractor (https://github.com/NixOS/nixpkgs/pull/123881)
- [x] nginxModules.vod (https://github.com/NixOS/nixpkgs/pull/142114)
- [x] nixos/modules/programs/ccache.nix
- [x] nwjs (https://github.com/NixOS/nixpkgs/pull/123796)
- [x] opencv2 (https://github.com/NixOS/nixpkgs/pull/123365)
- [x] opencv3 (https://github.com/NixOS/nixpkgs/pull/123365)
- [x] opencv4 (https://github.com/NixOS/nixpkgs/pull/123365)
- [x] openjfx11 (https://github.com/NixOS/nixpkgs/pull/123500)
- [x] openmw (#123607)
- [x] openrw (https://github.com/NixOS/nixpkgs/pull/141997)
- [x] openscenegraph (https://github.com/NixOS/nixpkgs/pull/121284)
- [x] oven-media-engine (https://github.com/NixOS/nixpkgs/pull/155537)
- [x] pangolin (https://github.com/NixOS/nixpkgs/pull/120735)
- [x] peek (https://github.com/NixOS/nixpkgs/pull/121812)
- [x] pianobar (https://github.com/NixOS/nixpkgs/pull/121486)
- [x] ppsspp (https://github.com/NixOS/nixpkgs/pull/120802)
- [x] pqiv (https://github.com/NixOS/nixpkgs/pull/121480)
- [x] privateer (https://github.com/NixOS/nixpkgs/pull/123483)
- [x] python2Packages.sipsimple (@pSub)
- [x] python2Packages.thumbor (https://github.com/NixOS/nixpkgs/pull/121257)
- [x] python3Packages.ha-ffmpeg (https://github.com/NixOS/nixpkgs/pull/121187)
- [x] python3Packages.imageio (https://github.com/NixOS/nixpkgs/pull/121257)
- [x] python3Packages.infoqscraper (https://github.com/NixOS/nixpkgs/pull/121257)
- [x] python3Packages.stytra (https://github.com/NixOS/nixpkgs/pull/121257)
- [x] qstopmotion (https://github.com/NixOS/nixpkgs/pull/123477)
- [x] renpy (https://github.com/NixOS/nixpkgs/pull/123464)
- [x] retroArchCores (https://github.com/NixOS/nixpkgs/pull/123842)
- [x] retroarchBare (https://github.com/NixOS/nixpkgs/pull/123514)
- [x] retroshare (https://github.com/NixOS/nixpkgs/pull/123345)
- [x] ring-daemon (@Radvendii @olynch)
- [x] spotify (https://github.com/NixOS/nixpkgs/pull/121468)
- [x] squeezelite (#122336)
- [x] togglesg-download (dropped)
- [x] tvheadend (https://github.com/NixOS/nixpkgs/pull/135662)
- [x] ultrastardx (https://github.com/NixOS/nixpkgs/pull/123416)
- [x] vdrPlugins.vaapidevice (https://github.com/NixOS/nixpkgs/pull/142141)
- [x] vivaldi (https://github.com/NixOS/nixpkgs/pull/123341)
- [x] wxSVG (https://github.com/NixOS/nixpkgs/pull/120802)
- [x] xineLib (https://github.com/NixOS/nixpkgs/pull/120802)
- [x] xscast (https://github.com/NixOS/nixpkgs/pull/123475)
- [x] yaxg (https://github.com/NixOS/nixpkgs/pull/123474)
- [x] yle-dl (https://github.com/NixOS/nixpkgs/pull/120773)
- [x] zoneminder (https://github.com/NixOS/nixpkgs/pull/123343)
- [x] amarok (https://github.com/NixOS/nixpkgs/pull/120657)
- [x] cantata (https://github.com/NixOS/nixpkgs/pull/120837)
- [x] carla (https://github.com/NixOS/nixpkgs/pull/121076)
- [x] ffmpeg-normalize (https://github.com/NixOS/nixpkgs/pull/120848)
- [x] get_iplayer (https://github.com/NixOS/nixpkgs/pull/120863)
- [x] libsForQt5.kfilemetadata (https://github.com/NixOS/nixpkgs/pull/120842)
- [x] mplayer (https://github.com/NixOS/nixpkgs/pull/120643)
- [x] python3Packages.ffmpeg-python (https://github.com/NixOS/nixpkgs/pull/120782)
- [x] qmmp (#120771)
- [x] qtox (https://github.com/NixOS/nixpkgs/pull/120839)
- [x] r128gain (https://github.com/NixOS/nixpkgs/pull/120783)
- [x] ripgrep-all (https://github.com/NixOS/nixpkgs/pull/120831)
- [x] streamlink (https://github.com/NixOS/nixpkgs/pull/120601)
- [x] tor-browser-bundle-bin (https://github.com/NixOS/nixpkgs/pull/120653)

Please remove the list of maintainers from packages that are done because GitHub won't allow me to ping more than a certain number of people.

<details>
<summary>script that I used to generate this list</summary>

The manually obtained file `packages` contains one attribute per line.
```bash
pkgs=$(cat packages)
for pkg in $pkgs; do
    pings=$(nix eval "(with import ./. { }; lib.concatStringsSep \" \" (map (m: \"@\" + m.github) ($pkg.meta.maintainers or [ ])))" --raw)
    if [ -z "$pings" ]; then
        echo "- [ ] $pkg" >> packages-with-maintainers
    else
        echo "- [ ] $pkg ($pings)" >> packages-with-maintainers
    fi
done
```
</details>

**Comment by meutraa** [comment]
Here is a list of all packages using ffmpeg_3(_x) 

```
nixos/modules/programs/ccache.nix
pkgs/applications/audio/amarok/default.nix
pkgs/applications/audio/ardour/default.nix
pkgs/applications/audio/bitwig-studio/bitwig-studio1.nix
pkgs/applications/audio/bitwig-studio/bitwig-studio3.nix
pkgs/applications/audio/cantata/default.nix
pkgs/applications/audio/carla/default.nix
pkgs/applications/audio/deadbeef/default.nix
pkgs/applications/audio/dr14_tmeter/default.nix
pkgs/applications/audio/kid3/default.nix
pkgs/applications/audio/moc/default.nix
pkgs/applications/audio/musikcube/default.nix
pkgs/applications/audio/pianobar/default.nix
pkgs/applications/audio/qmmp/default.nix
pkgs/applications/audio/r128gain/default.nix
pkgs/applications/audio/spotify/default.nix
pkgs/applications/audio/squeezelite/default.nix
pkgs/applications/gis/grass/default.nix
pkgs/applications/graphics/pqiv/default.nix
pkgs/applications/kde/k3b.nix
pkgs/applications/misc/airtame/default.nix
pkgs/applications/misc/get_iplayer/default.nix
pkgs/applications/misc/goldendict/default.nix
pkgs/applications/misc/megacmd/default.nix
pkgs/applications/misc/megasync/default.nix
pkgs/applications/misc/moonlight-embedded/default.nix
pkgs/applications/misc/sigal/default.nix
pkgs/applications/networking/browsers/tor-browser-bundle-bin/default.nix
pkgs/applications/networking/browsers/vivaldi/default.nix
pkgs/applications/networking/instant-messengers/baresip/default.nix
pkgs/applications/networking/instant-messengers/linphone/default.nix
pkgs/applications/networking/instant-messengers/qtox/default.nix
pkgs/applications/networking/instant-messengers/ring-daemon/default.nix
pkgs/applications/networking/p2p/retroshare/default.nix
pkgs/applications/video/aegisub/default.nix
pkgs/applications/video/avxsynth/default.nix
pkgs/applications/video/bombono/default.nix
pkgs/applications/video/clipgrab/default.nix
pkgs/applications/video/devede/default.nix
pkgs/applications/video/dvd-slideshow/default.nix
pkgs/applications/video/ffmpeg-normalize/default.nix
pkgs/applications/video/gnomecast/default.nix
pkgs/applications/video/makemkv/default.nix
pkgs/applications/video/manim/default.nix
pkgs/applications/video/mplayer/default.nix
pkgs/applications/video/natron/default.nix
pkgs/applications/video/openshot-qt/libopenshot.nix
pkgs/applications/video/peek/default.nix
pkgs/applications/video/qstopmotion/default.nix
pkgs/applications/video/simplescreenrecorder/default.nix
pkgs/applications/video/streamlink/default.nix
pkgs/applications/video/vdr/plugins.nix
pkgs/applications/video/xscast/default.nix
pkgs/development/compilers/openjdk/openjfx/11.nix
pkgs/development/compilers/oraclejdk/jdk-linux-base.nix
pkgs/development/haskell-modules/configuration-nix.nix
pkgs/development/interpreters/renpy/default.nix
pkgs/development/libraries/ffmpegthumbnailer/default.nix
pkgs/development/libraries/ffms/default.nix
pkgs/development/libraries/kde-frameworks/kfilemetadata/default.nix
pkgs/development/libraries/libextractor/default.nix
pkgs/development/libraries/libgroove/default.nix
pkgs/development/libraries/liblinphone/default.nix
pkgs/development/libraries/libvdpau-va-gl/default.nix
pkgs/development/libraries/mediastreamer/default.nix
pkgs/development/libraries/opencv/3.x.nix
pkgs/development/libraries/opencv/4.x.nix
pkgs/development/libraries/opencv/default.nix
pkgs/development/libraries/openscenegraph/default.nix
pkgs/development/libraries/pangolin/default.nix
pkgs/development/libraries/qt-5/modules/qtwebengine.nix
pkgs/development/libraries/wxSVG/default.nix
pkgs/development/libraries/xine-lib/default.nix
pkgs/development/python-modules/ffmpeg-python/default.nix
pkgs/development/python-modules/ha-ffmpeg/default.nix
pkgs/development/python-modules/imageio/default.nix
pkgs/development/python-modules/infoqscraper/default.nix
pkgs/development/python-modules/sipsimple/default.nix
pkgs/development/python-modules/stytra/default.nix
pkgs/development/python-modules/thumbor/default.nix
pkgs/development/tools/nwjs/default.nix
pkgs/games/hedgewars/default.nix
pkgs/games/openmw/default.nix
pkgs/games/openrw/default.nix
pkgs/games/privateer/default.nix
pkgs/misc/emulators/attract-mode/default.nix
pkgs/misc/emulators/mgba/default.nix
pkgs/misc/emulators/ppsspp/default.nix
pkgs/misc/emulators/retroarch/cores.nix
pkgs/misc/emulators/retroarch/default.nix
pkgs/misc/lightspark/default.nix
pkgs/servers/http/nginx/modules.nix
pkgs/servers/mediatomb/default.nix
pkgs/servers/misc/oven-media-engine/default.nix
pkgs/servers/tvheadend/default.nix
pkgs/servers/zoneminder/default.nix
pkgs/tools/graphics/cfdg/default.nix
pkgs/tools/graphics/yaxg/default.nix
pkgs/tools/misc/capture/default.nix
pkgs/tools/misc/togglesg-download/default.nix
pkgs/tools/misc/yle-dl/default.nix
pkgs/tools/networking/minidlna/default.nix
pkgs/tools/text/ripgrep-all/default.nix
pkgs/tools/video/gopro/default.nix
```

**Comment by Radvendii** [comment]
Ring daemon is already broken. You're welcome to switch the ffmpeg version, and if I ever get around to fixing that package, I'll make sure it works (or downgrade the ffmpeg version if absolutely necessary)

**Comment by dotlambda** [comment]
> Ring daemon is already broken. You're welcome to switch the ffmpeg version, and if I ever get around to fixing that package, I'll make sure it works (or downgrade the ffmpeg version if absolutely necessary)

If a package is broken for a long time, we should consider removing it. But given that you want to fix it at some point, I guess we should keep this one.

> I can do it myself, but I'm assuming it's better to do one large PR changing as many as possible at once, than a million tiny PRs. Let me know if that's an incorrect assumption, and I should bump the ffmpeg version in ring-daemon

It's easier to do in small PRs because nobody will be able to test whether all these packages still work after switching to a newer Ffmpeg.

**Comment by magnetophon** [comment]
Ardour builds and runs with regular ffmpeg.
A new release is expected any day now, is it OK if I fix it in the PR for that?

**Comment by andrewrk** [comment]
FYI I'm working on libgroove upstream to bump the dependency. If we want to remove libgroove from nix and resubmit it later when I've completed that work that would be fine with me. Afaik the only package that uses it is groove basin, which I am also the upstream author of and also in the (long) process of rebooting it. IMO these packages should be removed from nixpkgs for now and I will resubmit them in the future.

**Comment by AndersonTorres** [comment]
> Ardour builds and runs with regular ffmpeg.
> A new release is expected any day now, is it OK if I fix it in the PR for that?

If it is a security update, then it is better to fix it now and backpropagate to current stable releases.

**Comment by AndersonTorres** [comment]
> FYI I'm working on libgroove upstream to bump the dependency. If we want to remove libgroove from nix and resubmit it later when I've completed that work that would be fine with me. Afaik the only package that uses it is groove basin, which I am also the upstream author of and also in the (long) process of rebooting it. IMO these packages should be removed from nixpkgs for now and I will resubmit them in the future.

Is this a distant future? If it is broken but you are notifying us that it will be fixed, therefore I would suggest mark `meta.broken = true;` followed with a comment `upstream notified a new version is coming` or something like that.

**Comment by AndersonTorres** [comment]
A question: why do you changed all ffmpeg references to ffmpeg_3 ones, if it was way easier and clearer to do it via `ffmpeg = ffmpeg_3;` in the relevant `callPackage`s?

**Comment by NekomimiScience** [comment]
I cannot submit a PR in a reasonable amount of time due to my current employer, but I can confirm that get_iplayer version `3.27` (SHA-256: `077y31gg020wjpx5pcivqgkqawcjxh5kjnvq97x2gd7i3wwc30qi`) builds and smoke tests fine using the ffmpeg package.

If someone would like to submit a PR updating the version and making the ffmpeg change, I would appreciate it greatly. ❤️

**Comment by minijackson** [comment]
Opened an issue for Carla here: https://github.com/falkTX/Carla/issues/1403

**Comment by peterhoeg** [comment]
I took the liberty of sorting the list to make it easier to read. I don't suggest we keep doing that - it was simply to get the the ones not done yet in order.

**Comment by jonringer** [comment]
thanks @dotlambda for organizing this, and tackling a good issue :)

**Comment by jonringer** [comment]
cc @NixOS/release-engineers this a great issue to address, and would be awesome to have completed before branch off :)

**Comment by dotlambda** [comment]
@jonringer If this is a blocker for 21.05, so should be https://github.com/NixOS/nixpkgs/pull/120582

**Comment by jonringer** [comment]
> @jonringer If this is a blocker for 21.05, so should be #120582

Agreed, we should avoid users having vulnerable software on their systems.

**Comment by sheenobu** [comment]
Testing spotify on 20.09 by merging the latest spotify into my .nixpkgs. had to use gnutls but I guess that's just a 20.09 specific problem? Commit is here: https://hg.sr.ht/~sheenobu/config/rev/98765d83a39647270a4f6e13d6f806400558f477

nixpkgs commit incoming, thanks!

**Comment by Profpatsch** [comment]
Ultrastardx ffmpeg 2.x -> 4.2: https://github.com/NixOS/nixpkgs/pull/123416

**Comment by dotlambda** [comment]
The remaining packages will be marked as insecure by https://github.com/NixOS/nixpkgs/pull/123496. So let's make a joint effort to bring the number of packages still using FFmpeg 3.4 as close to zero as possible before 21.05 is released!

**Comment by badmutex** [comment]
Vivaldi can be ticked

**Comment by lunik1** [comment]
MEGAcmd will need a patch or upstream fix to build against ffmpeg 4.4, as it accesses fields that were made private between 4.3 and 4.4 (see #122405 and FFmpeg/FFmpeg@108864a).

The problem is in the mega sdk so MEGASync might be affected. Looks like @michojel is the maintainer, can you check if this is the case? ffmpeg 4.4 is avaialbe on `staging-next`.

Both still build against ffmpeg 4.4. Are we targeting 4.3 or 4.4 for 21.05?

**Comment by dotlambda** [comment]
> Both still build against ffmpeg 4.4. Are we targeting 4.3 or 4.4 for 21.05?

staging-next will be merged into master before branch-off, so 21.05 will have FFmpeg 4.4

**Comment by michojel** [comment]
> The problem is in the mega sdk so MEGASync might be affected. Looks like @michojel is the maintainer, can you check if this is the case? ffmpeg 4.4 is available on `staging-next`.

It does fail on staging-next against ffmpeg 4.4:
```
mega/src/gfx/qt.cpp:1218:18: error: 'AVStream' {aka 'struct AVStream'} has no member named 'skip_to_keyframe'
 1218 |     videoStream->skip_to_keyframe = true;
      |                  ^~~~~~~~~~~~~~~~
```

I don't have time to work on a patch until next week. But since ffmpeg is optional, shall I temporarily remove the dependency?

**Comment by lunik1** [comment]
I'm looking into `retroArchCores`. Seems the only core that uses ffmpeg is PPSSPP which has already been ported, so shouldn't be too much of a challenge.

**Comment by lunik1** [comment]
I'm cursed. PPSSPP is affected by the same issue as the MEGA SDK - build broken by FFmpeg/FFmpeg@108864a.

The good news is that in this case upstream has fixed it, just not in the latest release. The PR fixing it, hrydgard/ppsspp#14176, applies cleanly to to the latest release. I have made PRs addressing this for the retroarch core and PPSSPP standalone at #123842 and #123843.

**Comment by zaninime** [comment]
I've tried updating nginxModules.vod but fails due to the usage of deprecated functions, an issue is open upstream: https://github.com/kaltura/nginx-vod-module/issues/1262

A similar case is there for thumbextractor (although I didn't contribute it), but it can be fixed by updating to the latest revision on master: https://github.com/NixOS/nixpkgs/pull/123881

**Comment by magnetophon** [comment]
Ardour fixed in the above PR.

**Comment by lunik1** [comment]
`retroArchCores` is done.

**Comment by loewenheim** [comment]
`pcsxr` is currently instantiated with `ffmpeg_2`, which is also marked as insecure.

**Comment by jchv** [comment]
Lightspark should be good as of #124330 and #124701.

**Comment by samuela** [comment]
It looks like PRs for capture and mediatomb have been merged now.

**Comment by eduardosm** [comment]
dr14_tmeter can be marked as done (https://github.com/NixOS/nixpkgs/pull/125241)

**Comment by hrdinka** [comment]
attract-mode was fixed as well in https://github.com/NixOS/nixpkgs/pull/126354

**Comment by flexagoon** [comment]
tvheadend was already fixed in #135662

**Comment by thiagokokada** [comment]
I went ahead and merged all remaining PRs that were already opened pointing on this issue.

**Comment by vs49688** [comment]
Related:
https://github.com/NixOS/nixpkgs/pull/155537
https://github.com/NixOS/nixpkgs/pull/155539

**Comment by ajs124** [comment]
After #155993 only grass and natron will be left, for both of which exist open PRs (#150286 and #121212 (nice)), but they sadly both seem to be stalled.

**Comment by jonringer** [comment]
grass has been merged

**Comment by SuperSandro2000** [comment]
and we are done here. Thank you everyone!

**Comment by dotlambda** [comment]
> and we are done here

What about natron?

**Comment by dotlambda** [comment]
Once this is actually done, ffmpeg_3 should be removed from Nixpkgs.

**Comment by SuperSandro2000** [comment]
We did that in #163509 I just forgot to tick the box,
